nav2_core/include/nav2_core/task_executor_at_waypoint_arrival.hpp

```diff
@@ -0,0 +1,88 @@
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef NAV2_CORE__TASK_EXECUTOR_AT_WAYPOINT_ARRIVAL_HPP_
+#define NAV2_CORE__TASK_EXECUTOR_AT_WAYPOINT_ARRIVAL_HPP_
+#pragma once
+
+#include <string>
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp_lifecycle/lifecycle_node.hpp"
+#include "geometry_msgs/msg/pose_stamped.hpp"
+
+namespace nav2_core
+{
+/**
+ * @brief Base class for creating a plugin in order to perform a specific task at waypoint arrivals.
+ *
+ */
+class TaskExecutorAtWaypointArrival
+{
+public:
+/**
+ * @brief Construct a new Simple Task Execution At Waypoint Base object
+ *
+ */
+  TaskExecutorAtWaypointArrival() {}
+
+  /**
+   * @brief Destroy the Simple Task Execution At Waypoint Base object
+   *
+   */
+  virtual ~TaskExecutorAtWaypointArrival() {}
+
+  /**
+   * @brief Override this to setup your pub, sub or any ros services that you will use in the plugin.
+   *
+   * @param parent parent node that plugin will be created within(for an example see nav_waypoint_follower)
+   * @param plugin_name plugin name comes from parameters in yaml file
+   */
+  virtual void initialize(
+    const rclcpp_lifecycle::LifecycleNode::WeakPtr & parent,
+    const std::string & plugin_name) = 0;
+
+  /**
+   * @brief Override this to define the body of your task that you would like to execute once the robot arrived to waypoint
+   *
+   * @param parent parent node that plugin will be created withing(waypoint_follower in this case)
+   * @param curr_pose current pose of the robot
+   * @param curr_waypoint_index current waypoint, that robot just arrived
+   */
+  virtual void processAtWaypoint(
+    const rclcpp_lifecycle::LifecycleNode::WeakPtr & parent,
```

Comment:
You don't need to give the parent again, they should have stored it if they wanted it

Reply:
Make the other 2 things const references